vortex-buffer/src/bit/buf.rs

```diff
@@ -0,0 +1,215 @@
+            buffer,
+            len,
+            offset: 0,
+        }
+    }
+
+    /// Create a new `BoolBuffer` backed by a [`ByteBuffer`] with `len` bits in view, starting at the
+    /// given `offset` (in bits).
+    ///
+    /// Panics if the buffer is not large enough to hold `len` bits or if the offset is greater than
+    pub fn new_with_offset(buffer: ByteBuffer, len: usize, offset: usize) -> Self {
+        assert!(
+            offset < len,
+            "cannot construct new BoolBuffer with offset {offset} len {len}"
+        );
+        assert!(
+            buffer.len() * 8 >= len - offset,
+            "provided ByteBuffer not large enough to back BoolBuffer with offset {offset} len {len}"
+        );
+
+        Self {
+            buffer,
+            len,
+            offset,
+        }
+    }
+
+    /// Create a new `BoolBuffer` of length `len` where all bits are set (true).
+    pub fn new_set(len: usize) -> Self {
+        let words = len.div_ceil(8);
+        let buffer = buffer![0xFF; words];
+
+        Self {
+            buffer,
+            len,
+            offset: 0,
+        }
+    }
+
+    /// Create a new `BoolBuffer` of length `len` where all bits are unset (false).
+    pub fn new_unset(len: usize) -> Self {
+        let words = len.div_ceil(8);
+        let buffer = buffer![0u8; words];
+
+        Self {
+            buffer,
+            len,
+            offset: 0,
+        }
+    }
+
+    /// Get the logical length of this `BoolBuffer`.
+    ///
+    /// This may differ from the physical length of the backing buffer, for example if it was
+    /// created using the `new_with_offset` constructor, or if it was sliced.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if the `BoolBuffer` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Offset of start of the buffer in bits.
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
+
+    /// Retrieve the value at the given index.
+    ///
+    /// Panics if the index is out of bounds.
+    pub fn value(&self, index: usize) -> bool {
+        assert!(index < self.len, "index {index} exceeds len {}", self.len);
+
+        let word_index = (self.offset + index) / 8;
+        let bit_index = (self.offset + index) % 8;
+        let word = self.buffer[word_index];
+        let bit = word & (1 << bit_index);
+
+        bit != 0
+    }
+
+    /// Create a new zero-copy slice of this BoolBuffer that begins at the `start` index and extends
+    /// for `len` bits.
+    ///
+    /// Panics if the slice would extend beyond the end of the buffer.
+    pub fn slice(&self, start: usize, len: usize) -> Self {
+        assert!(
+            start + len <= self.len,
+            "slice of len {len} starting at {start} exceeds len {}",
+            self.len
+        );
+
+        Self {
+            buffer: self.buffer.clone(),
+            len,
+            offset: self.offset + start,
```

Comment:
Do we want to guarantee that we only hold a bit_offset? i.e. `offset < 8`?

Reply:
Hmm, does that help much?